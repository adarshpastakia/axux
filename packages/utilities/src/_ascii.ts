/**
 * AxUX React UI Framework with Pure CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */
/* istanbul ignore file */

import { isEmpty } from "./_isType";

/** ***************** latin-ascii conversions *******************/
const conversions = {
  AE: "Æ|Ǽ",
  Ae: "Ä",
  ae: "ä|æ|ǽ",
  OE: "Œ",
  Oe: "Ö",
  oe: "ö|œ",
  Ue: "Ü",
  ue: "ü",
  IJ: "Ĳ",
  ij: "ĳ",
  ss: "ß",
  A: "À|Á|Â|Ã|Ä|Å|Ǻ|Ā|Ă|Ą|Ǎ",
  a: "à|á|â|ã|å|ǻ|ā|ă|ą|ǎ|ª",
  C: "Ç|Ć|Ĉ|Ċ|Č",
  c: "ç|ć|ĉ|ċ|č",
  D: "Ð|Ď|Đ",
  d: "ð|ď|đ",
  E: "È|É|Ê|Ë|Ē|Ĕ|Ė|Ę|Ě",
  e: "è|é|ê|ë|ē|ĕ|ė|ę|ě",
  G: "Ĝ|Ğ|Ġ|Ģ",
  g: "ĝ|ğ|ġ|ģ",
  H: "Ĥ|Ħ",
  h: "ĥ|ħ",
  I: "Ì|Í|Î|Ï|Ĩ|Ī|Ĭ|Ǐ|Į|İ",
  i: "ì|í|î|ï|ĩ|ī|ĭ|ǐ|į|ı",
  J: "Ĵ",
  j: "ĵ",
  K: "Ķ",
  k: "ķ",
  L: "Ĺ|Ļ|Ľ|Ŀ|Ł",
  l: "ĺ|ļ|ľ|ŀ|ł",
  N: "Ñ|Ń|Ņ|Ň",
  n: "ñ|ń|ņ|ň|ŉ",
  O: "Ò|Ó|Ô|Õ|Ō|Ŏ|Ǒ|Ő|Ơ|Ø|Ǿ",
  o: "ò|ó|ô|õ|ō|ŏ|ǒ|ő|ơ|ø|ǿ|º",
  R: "Ŕ|Ŗ|Ř",
  r: "ŕ|ŗ|ř",
  S: "Ś|Ŝ|Ş|Š",
  s: "ś|ŝ|ş|š|ſ",
  T: "Ţ|Ť|Ŧ",
  t: "ţ|ť|ŧ",
  U: "Ù|Ú|Û|Ũ|Ū|Ŭ|Ů|Ű|Ų|Ư|Ǔ|Ǖ|Ǘ|Ǚ|Ǜ",
  u: "ù|ú|û|ũ|ū|ŭ|ů|ű|ų|ư|ǔ|ǖ|ǘ|ǚ|ǜ",
  Y: "Ý|Ÿ|Ŷ",
  y: "ý|ÿ|ŷ",
  W: "Ŵ",
  w: "ŵ",
  Z: "Ź|Ż|Ž",
  z: "ź|ż|ž",
  f: "ƒ",
};

/**
 * latin to ascii converter for convenient searching
 * @internal
 */
export const ascii = (str: string) => {
  if (isEmpty(str)) return "";
  Object.entries(conversions).forEach(([a, l]) => {
    const re = new RegExp(l, "g");
    str = `${str}`.replace(re, a);
  });
  return str;
};
