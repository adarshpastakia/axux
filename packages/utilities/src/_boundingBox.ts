/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

const _getBox = (box: AnyObject = "0,0,0,0") => {
  let boundingBox = box;
  if (typeof box === "string") {
    boundingBox = box.split(",");
  }
  if (Array.isArray(boundingBox)) {
    boundingBox = boundingBox.map((b) => parseInt(b, 10));
  }
  return boundingBox;
};

/**
 * box "x,y,w,h" | [x,y,w,h]
 * returns canvas box
 */
export const getBoundingBox = (box: AnyObject = "0,0,0,0") => {
  const boundingBox = _getBox(box);

  let [x = 0, y = 0, w = 0, h = 0] = boundingBox;
  x = x > 0 ? x : 0;
  y = y > 0 ? y : 0;
  w = w - x;
  h = h - y;
  return [x, y, w, h] as [number, number, number, number];
};

/**
 * box "x1,y1,x2,y2" | [x1,y1,x2,y2]
 * returns canvas box
 */
export const getBox = (box: AnyObject = "0,0,0,0") => {
  const boundingBox = _getBox(box);

  let [x1 = 0, y1 = 0, x2 = 0, y2 = 0] = boundingBox;
  x1 = x1 > 0 ? x1 : 0;
  y1 = y1 > 0 ? y1 : 0;
  return [x1, y1, x2, y2] as [number, number, number, number];
};
