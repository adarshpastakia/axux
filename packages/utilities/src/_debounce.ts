/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */
/* istanbul ignore file */

import { hash } from "./_hash";

const _debounce = {
  _timer: {} as KeyValue,
  fn: (fn: (...rest: AnyObject) => AnyObject, timeout = 500) => {
    /******************* create debounce timer id using callback hash *******************/
    const id = hash(fn.toString());

    /******************* return debounced function *******************/
    const cb = (...args: AnyObject) => {
      _debounce._timer[id] && clearTimeout(_debounce._timer[id]);
      _debounce._timer[id] = window.setTimeout(() => fn(...args), timeout);
    };
    cb.cancel = () => clearTimeout(_debounce._timer[id]);
    return cb;
  },
};

/**
 * debounce callback
 * @internal
 */
export const debounce = _debounce.fn;
