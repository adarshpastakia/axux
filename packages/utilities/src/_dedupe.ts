/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

import { isEmpty } from "./_isType";

export const dedupe = (list: AnyObject[], key?: string) => {
  const ret = Object.values(
    list.reduce<KeyValue>(
      (r, c) => ({ ...r, [key ? c[key] : c]: c }),
      {}
    )
  );
  return ret.filter(Boolean);
};

export const flatten = (list: AnyObject[]) => {
  return list.flat(5).filter((i) => !isEmpty(i));
};

export const flattenAndDedupe = (list: AnyObject[] = [], key?: string) => {
  return dedupe(flatten(list), key);
};
