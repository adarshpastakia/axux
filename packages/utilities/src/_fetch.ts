/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */
/* istanbul ignore file */

import { isString } from "./_isType";

declare type METHODS = "get" | "post" | "put" | "delete";
declare type FetchOptions = {
  body?: KeyValue | FormData;
  headers?: KeyValue;
  signal?: string | AbortSignal;
};

const _signals: Map<string, AbortController> = new Map();

/**
 * Fetch helper
 * @param method
 * @param url
 * @param headers
 * @param body
 * @param signal
 * @internal
 */
export const _fetch = (
  method: METHODS,
  url: string,
  { headers = {}, body, signal }: FetchOptions
) => {
  const options: KeyValue = {
    method,
  };

  headers["accept"] = "application/json";
  if (method === "post" || method === "put") {
    options.body = body;
    /******************* set body to json when not FormData *******************/
    if (!(body instanceof FormData)) {
      headers["content-type"] = "application/json";
      options.body = JSON.stringify(body);
    }
  }

  /******************* reset abort signal *******************/
  if (signal && isString(signal)) {
    if (_signals.has(signal)) {
      const _s = _signals.get(signal) as AbortController;
      _s.abort();
      _signals.delete(signal);
    }
    const ac = new AbortController();
    options.signal = ac.signal;
    _signals.set(signal, ac);
  } else if (signal instanceof AbortSignal) {
    options.signal = signal;
  }

  return fetch(url, { ...options, headers })
    .then((resp) => {
      if (resp.status >= 200 && resp.status < 300) {
        return resp.json();
      }
      throw { status: resp.status, code: -1, message: resp.statusText };
    })
    .then((resp) => {
      /******************* check for possible error code propeties in response *******************/
      const code = resp.error_code ?? resp.errorCode ?? resp.error;
      const message = resp.error_message ?? resp.errorMessage ?? resp.message;
      if (resp.status === "error" || code) {
        throw { status: 500, code, message };
      }
      return resp;
    });
};

/******************* attach isAborted helper to fetch method *******************/
_fetch.isAborted = ({ name }: Error) => name === "AbortError";

/******************* attach abort helper to fetch method *******************/
_fetch.abort = (signal: string) => {
  if (_signals.has(signal)) {
    const _s = _signals.get(signal) as AbortController;
    _s.abort();
    _signals.delete(signal);
  }
};
