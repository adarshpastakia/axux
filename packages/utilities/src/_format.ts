// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { differenceInDays, format, parseISO } from "date-fns";
import { CountryCode, parsePhoneNumberFromString } from "libphonenumber-js";
import numeral from "numeral";
import { isEmpty, isNil } from "./_isType";
import { Countries } from "./index";

export namespace Format {
  const getPhone = (value = "", country = "ae") => {
    const phone = value.match(/^[0+]/) ? value.replace(/^00/, "+") : `+${value}`;
    const number = parsePhoneNumberFromString(phone, country as CountryCode);
    return number
      ? number
      : {
          country: "",
          formatNational: () => value,
          formatInternational: () => value
        };
  };

  const numberFormat = (number?: string | number, format = "0,0[.]00a") => {
    if (isNil(number)) {
      return "";
    }
    const prefix = `${number}`.startsWith("+") ? "+" : "";
    return prefix + numeral(number).format(format);
  };

  export const phone = (value?: string) => {
    if (isEmpty(value) || !/^[\d+\s\-()]+$/.test(value)) return undefined;
    const phone = getPhone(value);
    return `${Countries.emoji(phone.country ?? "")} ${phone.formatInternational()}`;
  };
  export const number = (number?: string | number, format?: string) => {
    return numberFormat(number, format);
  };
  export const bytes = (number?: string | number) => {
    return numberFormat(number, "0,0[.]00b");
  };
  export const percent = (number?: string | number) => {
    return numberFormat(number, "0,0[.]00%");
  };
  export const duration = (number?: string | number, isFraction?: boolean) => {
    if (isNil(number)) {
      return "00:00.000";
    }
    const value = parseFloat(`${number}`) * (isFraction ? 1000 : 1);
    if (!isNaN(value)) {
      const days = differenceInDays(value, 0);
      const prefix = days > 0 ? `${days}d:` : "";
      let time = format(parseISO(new Date(value).toISOString().replace("Z", "")), "HH:mm:ss.SSS");
      if (!prefix) time = time.replace(/^00:/, "");
      return `${prefix}${time}`;
    }
    return "00:00.000";
  };
}
