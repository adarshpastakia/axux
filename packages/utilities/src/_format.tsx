/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

import { differenceInDays, format, parseISO } from "date-fns";
import { CountryCode, parsePhoneNumberFromString } from "libphonenumber-js";
import numeral from "numeral";
import { isEmpty, isNil } from "./_isType";

export namespace Format {
  /******************* get phone number object using `libphonenumber-js` *******************/
  const getPhone = (value = "", country = "ae") => {
    const phone = value.match(/^[0+]/)
      ? value.replace(/^00/, "+")
      : `+${value}`;
    const number = parsePhoneNumberFromString(phone, country as CountryCode);
    return number
      ? number
      : {
          country: "",
          formatNational: () => value,
          formatInternational: () => value,
        };
  };

  /******************* common number format method using `numeral` *******************/
  const numberFormat = (number?: string | number, format = "0,0[.]00a") => {
    if (isNil(number)) {
      return "";
    }
    const prefix = `${number}`.startsWith("+") ? "+" : "";
    return prefix + numeral(number).format(format);
  };

  const makeDuration = (value: number, fmt = "HH:mm:ss.SSS") => {
    if (!isNaN(value)) {
      const days = differenceInDays(value, 0);
      const prefix = days > 0 ? `${days}d:` : "";
      let time = format(
        parseISO(new Date(value).toISOString().replace("Z", "")),
        fmt
      );
      if (!prefix) time = time.replace(/^00:/, "");
      return `${prefix}${time}`;
    }
  };

  /******************* format phone number using `libphonenumber-js` *******************/
  export const phone = (value?: string) => {
    if (isEmpty(value) || !/^[\d+\s\-()]+$/.test(value)) return undefined;
    const phone = getPhone(value);
    return (
      <>
        <span className={` flag ${phone.country ?? ""}`} />
        &nbsp;<span>{phone.formatInternational()}</span>
      </>
    );
  };
  /******************* format whole number using `numeral` *******************/
  export const number = (number?: string | number, format?: string) => {
    return numberFormat(number, format);
  };
  /******************* format bytes using `numeral` *******************/
  export const bytes = (number?: string | number) => {
    return numberFormat(number, "0,0[.]00b");
  };
  /******************* format percentage using `numeral` *******************/
  export const percent = (number?: string | number) => {
    return numberFormat(number, "0,0[.]00%");
  };
  /******************* format date using `date-fns` *******************/
  export const date = (date?: Date | string | number, fmt = "d-M-y") => {
    return date ? format(new Date(date), fmt) : "";
  };
  /******************* format time duration from total seconds/milliseconds *******************/
  export const duration = (number?: string | number, isFraction?: boolean) => {
    if (isNil(number)) {
      return "00:00.000";
    }
    const value = parseFloat(`${number}`) * (isFraction ? 1000 : 1);
    if (!isNaN(value)) {
      return makeDuration(value, "HH:mm:ss.SSS");
    }
    return "00:00.000";
  };
  export const durationSeconds = (
    number?: string | number,
    isFraction?: boolean
  ) => {
    if (isNil(number)) {
      return "00:00";
    }
    const value = parseFloat(`${number}`) * (isFraction ? 1000 : 1);
    if (!isNaN(value)) {
      return makeDuration(value, "HH:mm:ss");
    }
    return "00:00";
  };
}
