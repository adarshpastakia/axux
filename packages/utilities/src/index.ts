/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

/* istanbul ignore file */

export {
  useEffectDebugger,
  useLayoutEffectDebugger,
} from "./useEffectDebugger";
export { useLogger } from "./useLogger";
export * from "./_ascii";
export { default as Countries } from "./_countries";
export type { Country } from "./_countries";
export * from "./_debounce";
export { _fetch as fetch } from "./_fetch";
export * from "./_format";
export * from "./_getByPath";
export * from "./_hash";
export * from "./_interpolate";
export * from "./_isEqual";
export * from "./_isType";
export * from "./_tokenize";
export * from "./_dedupe";
export * from "./_boundingBox";

import { default as Countries } from "./_countries";

// @ts-ignore
window.Countries = Countries;
