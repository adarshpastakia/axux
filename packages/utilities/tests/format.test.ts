/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

import { Format } from "../src";

describe("format test", () => {
  it("should format number", (done) => {
    expect(Format.number()).toBe("");
    expect(Format.number(9.9)).toBe("9.90");
    expect(Format.number(9990)).toBe("9.99k");
    expect(Format.number("-9.81")).toBe("-9.81");
    expect(Format.number("+9.24")).toBe("+9.24");
    done();
  });

  it("should format bytes", (done) => {
    expect(Format.bytes()).toBe("");
    expect(Format.bytes(1000)).toBe("1KB");
    expect(Format.bytes(1000 * 1000)).toBe("1MB");
    expect(Format.bytes(1000 * 1000 * 1000)).toBe("1GB");
    done();
  });

  it("should format percentage", (done) => {
    expect(Format.percent()).toBe("");
    expect(Format.percent(0.24)).toBe("24%");
    expect(Format.percent(0.2418)).toBe("24.18%");
    done();
  });

  it("should format date", (done) => {
    expect(Format.date()).toBe("");
    expect(Format.date("2020-01-01")).toBe("1-1-2020");
    expect(Format.date("2020-01-01", "dd-MMM-y")).toBe("01-Jan-2020");
    done();
  });

  it("should format duration", (done) => {
    expect(Format.duration()).toBe("00:00.000");
    expect(Format.duration("test")).toBe("00:00.000");
    expect(Format.duration(86400000 + 14400000 + 1740000 + 18000 + 189)).toBe(
      "1d:04:29:18.189"
    );
    expect(Format.duration(60 * 24 + 3 + 0.128, true)).toBe("24:03.128");
    done();
  });

  it("should format duration seconds", (done) => {
    expect(Format.durationSeconds()).toBe("00:00");
    expect(Format.durationSeconds("test")).toBe("00:00");
    expect(Format.durationSeconds(1000 * 60 * 24 + 3000 + 128)).toBe("24:03");
    expect(Format.durationSeconds(60 * 24 + 3 + 0.128, true)).toBe("24:03");
    done();
  });

  it("should format phone", (done) => {
    expect(Format.phone()).toBe(undefined);
    expect(Format.phone("99123")).toMatchSnapshot();
    expect(Format.phone("971502491824")).toMatchSnapshot();
    expect(Format.phone("00971502491824")).toMatchSnapshot();
    done();
  });
});
