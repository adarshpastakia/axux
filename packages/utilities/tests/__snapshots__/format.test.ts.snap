// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`format test should format phone 1`] = `"🏳️ 99123"`;

exports[`format test should format phone 2`] = `"🇦🇪 +971 50 249 1824"`;

exports[`format test should format phone 3`] = `"🇦🇪 +971 50 249 1824"`;
